File: manager/src/main/java/moe/shizuku/manager/viewmodel/AppsViewModel.java
package moe.shizuku.manager.viewmodel;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Single;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import moe.shizuku.manager.BuildConfig;
import moe.shizuku.manager.authorization.AuthorizationManager;
import moe.shizuku.manager.utils.AppNameComparator;

public class AppsViewModel extends SharedViewModel {

    private final MutableLiveData<DataWrapper<List<PackageInfo>>> mPackages = new MutableLiveData<>();
    private final MutableLiveData<DataWrapper<Integer>> mGrantedCount = new MutableLiveData<>();
    private Disposable mPackagesDisposable;
    private Disposable mCountDisposable;

    public MutableLiveData<DataWrapper<List<PackageInfo>>> getPackages() {
        return mPackages;
    }

    public MutableLiveData<DataWrapper<Integer>> getGrantedCount() {
        return mGrantedCount;
    }

    @Override
    protected void onFullyCleared() {
        if (mPackagesDisposable != null)
            mPackagesDisposable.dispose();

        if (mCountDisposable != null)
            mCountDisposable.dispose();
    }

    public void load(Context context) {
        mPackagesDisposable = Single
                .fromCallable(() -> {
                    List<PackageInfo> list = new ArrayList<>();
                    List<String> packages = new ArrayList<>();
                    for (PackageInfo pi : AuthorizationManager.getPackages(PackageManager.GET_META_DATA)) {
                        if (BuildConfig.APPLICATION_ID.equals(pi.packageName))
                            continue;

                        list.add(pi);

                        if (AuthorizationManager.granted(pi.packageName, pi.applicationInfo.uid))
                            packages.add(pi.packageName);
                    }
                    Collections.sort(list, new AppNameComparator(context).getPackageInfoComparator());
                    return new List[]{list, packages};
                })
                .subscribeOn(Schedulers.io())
                .subscribe(lists -> {
                    //noinspection unchecked
                    mPackages.postValue(new DataWrapper<>(lists[0]));
                    mGrantedCount.postValue(new DataWrapper<>(lists[1].size()));
                }, throwable -> {
                    mPackages.postValue(new DataWrapper<>(throwable));
                    mGrantedCount.postValue(new DataWrapper<>(throwable));
                });
    }

    public void loadCount() {
        mCountDisposable = Single
                .fromCallable(() -> {
                    int count = 0;
                    for (PackageInfo pi : AuthorizationManager.getPackages(0)) {
                        if (BuildConfig.APPLICATION_ID.equals(pi.packageName))
                            continue;

                        count += AuthorizationManager.granted(pi.packageName, pi.applicationInfo.uid) ? 1 : 0;
                    }
                    return count;
                })
                .subscribeOn(Schedulers.io())
                .subscribe(count -> mGrantedCount.postValue(new DataWrapper<>(count)), throwable -> {
                    mGrantedCount.postValue(new DataWrapper<>(throwable));
                });
    }
}

File: manager/src/main/java/moe/shizuku/manager/management/AppsViewModel.kt
package moe.shizuku.manager.management

import android.content.Context
import android.content.pm.PackageInfo
import androidx.activity.ComponentActivity
import androidx.annotation.MainThread
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import moe.shizuku.manager.authorization.AuthorizationManager
import rikka.lifecycle.Resource
import rikka.lifecycle.activityViewModels
import rikka.lifecycle.viewModels

@MainThread
fun ComponentActivity.appsViewModel() = viewModels { AppsViewModel(this) }

@MainThread
fun Fragment.appsViewModel() = activityViewModels { AppsViewModel(requireContext()) }

class AppsViewModel(context: Context) : ViewModel() {

    private val _packages = MutableLiveData<Resource<List<PackageInfo>>>()
    val packages = _packages as LiveData<Resource<List<PackageInfo>>>

    private val _grantedCount = MutableLiveData<Resource<Int>>()
    val grantedCount = _grantedCount as LiveData<Resource<Int>>

    fun load(onlyCount: Boolean = false) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val list: MutableList<PackageInfo> = ArrayList()
                var count = 0
                for (pi in AuthorizationManager.getPackages()) {
                    list.add(pi)
                    if (AuthorizationManager.granted(pi.packageName, pi.applicationInfo!!.uid)) count++
                }
                if (!onlyCount) _packages.postValue(Resource.success(list))
                _grantedCount.postValue(Resource.success(count))
            } catch (e: CancellationException) {

            } catch (e: Throwable) {
                _packages.postValue(Resource.error(e, null))
                _grantedCount.postValue(Resource.error(e, 0))
            }
        }
    }
}
